include/deal.II/lac/solver_gmres.h

```diff
@@ -710,81 +723,319 @@ SolverGMRES<VectorType>::givens_rotation(Vector<double> &h,
 
 
 
-template <class VectorType>
-inline double
-SolverGMRES<VectorType>::iterated_modified_gram_schmidt(
-  const internal::SolverGMRESImplementation::TmpVectors<VectorType>
-    &                                       orthogonal_vectors,
-  const unsigned int                        dim,
-  const unsigned int                        accumulated_iterations,
-  VectorType &                              vv,
-  Vector<double> &                          h,
-  bool &                                    reorthogonalize,
-  const boost::signals2::signal<void(int)> &reorthogonalize_signal)
+namespace internal
 {
-  Assert(dim > 0, ExcInternalError());
-  const unsigned int inner_iteration = dim - 1;
+  namespace SolverGMRESImplementation
+  {
+    template <class VectorType>
+    void
+    dot_and_add(
+      const unsigned int dim,
+      const VectorType & vv,
+      const internal::SolverGMRESImplementation::TmpVectors<VectorType>
+        &             orthogonal_vectors,
+      Vector<double> &h)
+    {
+      for (unsigned int i = 0; i < dim; ++i)
+        h[i] += vv * orthogonal_vectors[i];
+    }
 
-  // need initial norm for detection of re-orthogonalization, see below
-  double     norm_vv_start = 0;
-  const bool consider_reorthogonalize =
-    (reorthogonalize == false) && (inner_iteration % 5 == 4);
-  if (consider_reorthogonalize)
-    norm_vv_start = vv.l2_norm();
 
-  for (unsigned int i = 0; i < dim; ++i)
-    h[i] = 0;
 
-  for (unsigned int c = 0; c < 2; ++c) // 0: orthogonalize, 1: reorthogonalize
+    template <class Number>
+    void
+    dot_and_add(
+      const unsigned int                                                   dim,
+      const LinearAlgebra::distributed::Vector<Number, MemorySpace::Host> &vv,
+      const internal::SolverGMRESImplementation::TmpVectors<
+        LinearAlgebra::distributed::Vector<Number, MemorySpace::Host>>
+        &             orthogonal_vectors,
+      Vector<double> &h)
     {
-      // Orthogonalization
-      double htmp = vv * orthogonal_vectors[0];
-      h(0) += htmp;
+      AlignedVector<VectorizedArray<double>> hs(dim,
+                                                VectorizedArray<double>(0.0));
+
+      unsigned int j = 0;
+      for (unsigned int c = 0;
+           j < vv.locally_owned_size() / VectorizedArray<double>::size();
+           ++c, j += VectorizedArray<double>::size())
```

Comment:
Is this correct? I would intuitively say
```suggestion
      for (unsigned int c = 0;
           c < vv.locally_owned_size() / VectorizedArray<double>::size();
           ++c, j += VectorizedArray<double>::size())
```